### PR TITLE
Add support for sparse tensors with the JAX backend.

### PR DESCRIPTION
**Pull request by hertschuh**
Add support for sparse tensors with the JAX backend.

This is experimental since JAX support for sparse tensors itself is experimental.

This support has the following limitation:
- Only the BCOO format is fully supported and tested. Other formats are untested and only supported for some use cases.
- Not all combinations of `n_batch`, `n_sparse` and `n_dense` are supported. What is tested is:
  - all sparse (equivalent to `tf.SparseTensor`)
  - `n_sparse == 1` and the rest is dense (equivalent to `tf.IndexedSlices`)
- The the embedding layer / the `take` op is not optimized (densifies inputs).
- Gradients obtained from sparse inputs are not sparse.

Also:
- Fixed a bug with TensorFlow sparse support for `squeeze` when `axis` is negative.
- `run_layer_test` can now take a tuple of input shapes.

### CODE REVIEW

File: keras/backend/tests/compute_output_spec_test.py [hunk 1]
```diff
@@ -61,9 +61,11 @@ def single_arg_sparse_fn(x):
         ys = backend.compute_output_spec(single_arg_sparse_fn, x)
         for y in ys:
             self.assertEqual(y.shape, (None, 3, 3))
-            self.assertTrue(y.sparse)
+            # Sparseness inference is only available with TensorFlow.
```
Comment (fchollet) [inline keras/backend/tests/compute_output_spec_test.py:64]: What would it take to make it possible with JAX?

**Review by fchollet — COMMENTED** [review]
Thanks for the PR! Looks well-factored.

File: keras/backend/tests/compute_output_spec_test.py (hunk 1 above)
Comment (hertschuh) [reply keras/backend/tests/compute_output_spec_test.py:64]: `compute_output_spec` is based on `jax.ShapeDtypeStruct` and `jax.make_jaxpr`, neither of them support the concept of sparse and I don't see a way to add it.
But maybe there's a way to inspect the resulting `ClosedJaxpr`? Let me look into it.

Comment (hertschuh) [reply keras/backend/tests/compute_output_spec_test.py:64]: Oh, I think I found a way to do it.

File: keras/backend/jax/sparse.py [hunk 2]
```diff
@@ -0,0 +1,328 @@
...
+                    # x1 is sparse BCOO, x2 is scalar, apply func element-wise.
+                    return jax_sparse.BCOO(
+                        (func(x1.data, x2), x1.indices),
+                        shape=x1.shape,
+                        indices_sorted=x1.indices_sorted,
+                        unique_indices=x1.unique_indices,
+                    )
+                else:
+                    # x1 is sparse BCOO, x2 is dense.
+                    if not jax_utils.is_in_jax_tracing_scope(x2):
```
Comment (hertschuh) [inline keras/backend/jax/sparse.py:301]: This is not great... The alternative is to not care about whether the dense divisor has zeros or NaNs (btw, the jax sparsify implementation does not check for zeros and NaNs).

File: keras/backend/tests/compute_output_spec_test.py (hunk 1 above)
Comment (hertschuh) [reply keras/backend/tests/compute_output_spec_test.py]: Added.

**Review by fchollet — APPROVED** [review]
LGTM, thank you!
